src/helpers/buffers/unordered_receiver.rs

```diff
@@ -0,0 +1,502 @@
+    }
+}
+
+/// Saved unread data from a received chunk.
+#[derive(Default)]
+struct Spare {
+    buf: Vec<u8>,
+    offset: usize,
+}
+
+impl Spare {
+    /// Read a message from the buffer.  Returns `None` if there isn't enough data.
+    fn read<M: Message>(&mut self) -> Option<M> {
+        println!("Read {} bytes", M::Size::USIZE);
+        let end = self.offset + M::Size::USIZE;
+        if end <= self.buf.len() {
+            let m = M::deserialize(GenericArray::from_slice(&self.buf[self.offset..end]));
+            self.offset = end;
+            Some(m)
+        } else {
+            None
+        }
+    }
+
+    /// Replace the stored value with the given slice.
+    fn replace(&mut self, v: &[u8]) {
+        self.offset = 0;
+        self.buf.truncate(0);
+        self.buf.extend_from_slice(v);
+    }
+
+    /// Extend the buffer with new data.  
+    /// This returns a message if there is enough data.
+    /// This returns a value because it can be more efficient in cases where
+    /// received chunks don't align with messages.
+    fn extend<M: Message>(&mut self, v: &[u8]) -> Option<M> {
+        let sz = <M::Size as Unsigned>::USIZE;
+        let remainder = self.buf.len() - self.offset;
+        if remainder + v.len() < sz {
+            // Not enough data: save it.
+            // If we're working from the tail of a longer buffer, only retain the tail.
+            self.buf = self.buf.split_off(self.offset);
+            self.buf.extend_from_slice(v);
+            self.offset = 0;
+            return None;
+        }
+
+        let m = if remainder > 0 {
+            // Copy to the stack to join old and new data.
+            let needed = sz - remainder;
+            let mut tmp = GenericArray::<u8, M::Size>::default();
+            tmp[..remainder].copy_from_slice(&self.buf[self.offset..]);
+            tmp[remainder..].copy_from_slice(&v[..needed]);
+            self.replace(&v[needed..]);
+            M::deserialize(&tmp)
+        } else {
+            self.replace(&v[sz..]);
+            M::deserialize(GenericArray::from_slice(&v[..sz]))
+        };
+        Some(m)
+    }
+}
+
+pub struct OperatingState<S, C>
+where
+    S: Stream<Item = C>,
+    C: AsRef<[u8]>,
+{
+    /// The stream we're reading from.
+    stream: Pin<Box<S>>,
+    /// This tracks `Waker` instances from calls to `recv()` with indices that
+    /// aren't ready at the time of the call.  If the future is invoked prior
+    /// to the value being ready, the `Waker` is saved here.
+    wakers: Vec<Option<Waker>>,
+    /// If we ever find that a waker doesn't fit in `wakers`, this is where
+    /// they are stashed.  These are more than `c` items past the current
+    /// item into the future when registered (c = capacity or `wakers.len()`).
+    /// So we don't want to wake them frequently.  Instead, these are woken on
+    /// a fixed cadence of every `c/2` items.
+    ///
+    /// When polled again, any that were only a little past the capacity will
```

Comment:
is it worth considering a future optimization where we only poll overflow wakers when there is nothing to be polled in `wakers`?

Reply:
That might be better, the more I think on it.  I was concerned that we might have to walk the main store to ensure that things were fine, but we can count the number of wakers that are in there.  The other cost is that we might flush the extras too often, but I think that only happens after you finish reading the last item in a chunk in most cases.  With that design, any time you read an item out and there are no outstanding receivers, you will flush the overflow.  That will flush more often than this, probably, but it might be simpler to reason about.

Reply:
> I was concerned that we might have to walk the main store to ensure that things were fine, but we can count the number of wakers that are in there.

yea I was also thinking about maintaining a counter of active wakers inside `wakers`. that shouldn't require walking through the vector and counting